### ISSUE TITLE
Perf of intel-tensorflow vs default tensorflow on BERT

### ISSUE THREAD

**Issue by WilliamTambellini** [intel/mkl-dnn#540]
This is a call to confirm or not the perf of "intel-tensorflow" as described here
https://software.intel.com/en-us/articles/intel-optimization-for-tensorflow-installation-guide
vs the default/naive tensorflow as given in the official pip repo, for a SQuAD BERT inference : 
https://github.com/google-research/bert

-----------------------------------------------------------------------------

### Environment
* CPU make and model (try `lscpu`; if your `lscpu` does not list CPU flags,
  try running `cat /proc/cpuinfo | grep flags | sort -u`)
Model name:            Intel(R) Core(TM) i7-4600U CPU @ 2.10GHz
* OS version (`uname -a`)
CentOS 7
* Compiler version (`gcc --version`)
Whatever compiler used to build TF by the python pip maintainers. 

* MKLDNN version: hard to say : from what I see in
/usr/lib/python3.6/site-packages/tensorflow/include/external/mkl_dnn/include/
the mkldnn headers copyrights are still from 2018.
$ ll /usr/lib/python3.6/site-packages/tensorflow/include/external/mkl_dnn/include
-rw-r--r--. 1 root root   2376 Aug 14 16:50 mkldnn_debug.h
-rw-r--r--. 1 root root  75949 Aug 14 16:50 mkldnn.h
-rw-r--r--. 1 root root 142047 Aug 14 16:50 mkldnn.hpp
-rw-r--r--. 1 root root  48357 Aug 14 16:50 mkldnn_types.h 
$ head /usr/lib/python3.6/site-packages/tensorflow/include/external/mkl_dnn/include/mkldnn.h
 Copyright 2016-2018 Intel Corporation
and there is no mkldnn_version symbol in the tf lib : 
$ nm /usr/lib/python3.6/site-packages/tensorflow/libtensorflow_framework.so | grep mkldnn_v
0000000000abfb00 T mkldnn_verbose_set
0000000000ab57a0 T mkldnn_view_primitive_desc_create
0000000000abfa40 T _ZN6mkldnn4impl14mkldnn_verboseEv
0000000001882aa0 b _ZZN6mkldnn4impl14mkldnn_verboseEvE11initialized

### Steps to reproduce
- train/finetune BERT on the SQuAD task  
https://github.com/google-research/bert#squad-11

- run inference (a simple batch of 10 questions on 1 paragraph, run twice for warmup) with the default tensorflow as intalled with pip:

pip3 install tensorflow==1.13.1
$ OMP_NUM_THREADS=1 python3 bert_inference.py 
Found TensortFlow version  1.13.1
Intel MKLDNN enabled ?  False
2019-08-20 17:46:52.979731: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Prediction 1 run in:  3.614552957005799  secs
Prediction 2 run in:  3.1976517980219796  secs

- Uninstall tf and install intel-tensorflow :
pip3 install intel-tensorflow
and rerun the very same inference:

$ OMP_NUM_THREADS=1 KMP_BLOCKTIME=0 KMP_AFFINITY=granularity=fine,verbose,compact,1,0 python3 bert_inference.py 
Found TensortFlow version  1.13.1
Intel MKLDNN enabled ?  True
OMP: Info #212: KMP_AFFINITY: decoding x2APIC ids.
OMP: Info #210: KMP_AFFINITY: Affinity capable, using global cpuid leaf 11 info
OMP: Info #154: KMP_AFFINITY: Initial OS proc set respected: 0-3
OMP: Info #156: KMP_AFFINITY: 4 available OS procs
OMP: Info #157: KMP_AFFINITY: Uniform topology
OMP: Info #179: KMP_AFFINITY: 1 packages x 2 cores/pkg x 2 threads/core (2 total cores)
OMP: Info #214: KMP_AFFINITY: OS proc to physical thread map:
OMP: Info #171: KMP_AFFINITY: OS proc 0 maps to package 0 core 0 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 2 maps to package 0 core 0 thread 1 
OMP: Info #171: KMP_AFFINITY: OS proc 1 maps to package 0 core 1 thread 0 
OMP: Info #171: KMP_AFFINITY: OS proc 3 maps to package 0 core 1 thread 1 
OMP: Info #250: KMP_AFFINITY: pid 7205 tid 7223 thread 0 bound to OS proc set 0
OMP: Info #250: KMP_AFFINITY: pid 7205 tid 7223 thread 1 bound to OS proc set 1
Prediction 1 run in:  5.119017199991504  secs
Prediction 2 run in:  4.5127738649898674  secs

Note: add print('Intel MKLDNN enabled ? ', tf.pywrap_tensorflow.IsMklEnabled())

### Actual behavior
intel-tensorflow slower than default tensorflow

### Expected behavior
intel-tensorflow faster than default tensorflow

**Comment by rsdubtso** [comment]
Can you please try running with OMP_NUM_THREADS=2? The two threads reported by the OpenMP actually belong to different teams, last time I checked, and thus the computations run sequentially. At the same time the default configuration most likely runs with 4 intraop threads.

**Comment by WilliamTambellini** [comment]
Hello @rsdubtso 

$ OMP_NUM_THREADS=2 python3 new_bert_inference.py
Intel MKLDNN enabled ?  False
Prediction 1 run in:  1.9485731390013825  secs
Prediction 2 run in:  1.5992247800168116  secs

$ OMP_NUM_THREADS=2  python3 new_bert_inference.py 
Intel MKLDNN enabled ?  True
Prediction 1 run in:  4.002434733993141  secs
Prediction 2 run in:  2.594828120985767  secs

I have not set any intra/inter op para (config.intra/inter_op_parallelism_threads) so TF to use the default I guess.

**Comment by rsdubtso** [comment]
Thanks! Paging @agramesh1 from the Intel TensorFlow team. To answer the original question: yes, such cases when Vanilla TensorFlow is faster than MKL-DNN do exist, but I am not sure about this one.

**Comment by vpirogov** [comment]
@WilliamTambelini,

I'm not sure what TF threading defaults are. Do you have single thread (intra_ops=inter_ops=omp_num_threads=1) comparison to establish the baseline?

**Comment by agramesh1** [comment]
Tagging @TensorFlow-MKL

**Comment by agramesh1** [comment]
@Tensorflow-MKL

**Comment by WilliamTambellini** [comment]
Just confirming this bad speed is probably specific to transformer style neural networks : I ve just tested a LSTM RNN and I see a good (40%) speed up by using intel-tensorflow over (Google ?) 'tensorflow' (as found/downloaded by the official pip).
Perhaps related to 
https://github.com/intel/mkl-dnn/issues/525
and/or
https://github.com/intel/mkl-dnn/issues/391
?

**Comment by louie-tsai** [comment]
@WilliamTambellini,

We are trying to reproduce your issue on our machines. will keep you posted.

**Comment by louie-tsai** [comment]
@WilliamTambellini,

After training, we want to verify inference by using exact command from your post. 
However, we didn't see  bert_inference.py file. 
would you mind to share the file with us?

thanks

**Comment by WilliamTambellini** [comment]
Hi @louie-tsai  
I cannot legally share that code but I could summarize how to do inference based on : 
https://github.com/google-research/bert
1st make sure to have bert-tensorflow installed.
$ pip3.7 list | grep tensor
bert-tensorflow               1.0.1    
https://pypi.org/project/bert-tensorflow/

then 
import bert
from bert import tokenization
set the TF parallelism accordingly:
    config = tf.ConfigProto()
    config.intra_op_parallelism_threads = args.tf_intra_op_threads
    config.inter_op_parallelism_threads = args.tf_inter_op_threads
    tf.Session(config=config)

load a bert tokenizer from your vocab file:
tokenizer = tokenization.FullTokenizer(vocab_file=vocab_dir, do_lower_case=True)

load a predictor model:
import tf predictor
from tensorflow.contrib import predictor
predict_fn = predictor.from_saved_model(model)

Use a BertTensor() object to build the tensor for the paragraph and the question
Use the predict_fn on the tensors to get the bert_object.predictions.

**Comment by vpirogov** [comment]
@louie-tsai, @agramesh1, any updates on that?

**Comment by louie-tsai** [comment]
@vpirogov  we are implementing the inference workload this week, and try to have conclusion next week.  sorry for late response.

**Comment by yinghu5** [comment]
Hi @WilliamTambellini 
I'm trying to reproduce your problem. Basically, i can see the slow performance with intel-tensorflow 1.13.1 version comparing default TF with run_squad.py.  which use very earlier mkldnn version. 

Could you please please try latest TF version with latest MKL DNN version  (it is better to have one build with  higher version than MKLDNN 0.18.1 version and see if there is performance change with your bert_inference.py? 

Because, as i learned, there is one fix in latest MKL-DNN version to address the bert performance.  With internal version of TF , we see good performance improvement. i will try later and update you if any result.

**Comment by yinghu5** [comment]
Hi @WilliamTambellini , 

I had tried to rebuild the TF 1.14 version with latest MKL-DNN v.20.3 .  and change the code based on here the code were changed based on   :  https://github.com/google-research/bert/pull/876.   run_xx.py  and modeling.py .  We saw about  over 1.5x performance boost on my lab machine.

**Comment by shuo-zhao** [comment]
Hello @louie-tsai @agramesh1 @vpirogov 
I have met the same issue and I tried all the version Tensorflow-mkl exist(1.13.1, 1.14.0, 2.0.0), it not performence better than a normal cpu one.

My cpu is Intel(R) Core(TM) i7-8700U CPU @ 3.20GHz with 6-core 12 logic processor.
OS: windows10.

When I run a very single ResNet50, the tensorflow with mkldnn costs almost same time as a normal one, but uses only 20% of cpu.

I tried all the intra_op_parallelism_threads, inter_op_parallelism_threads & OMP_NUM_THREADS, but it always use less than 20% of cpu.

Could you pelease help me about whcih else to set to make tensorflow with mkldnn perforence better?

thanks a lot!

**Comment by yinghu5** [comment]
Hi Shuo-zhao, 

What is the exact intra_op_parallelism_threads, inter_op_parallelism_threads & OMP_NUM_THREADS did you try?  
Here is article for your reference
https://software.intel.com/en-us/articles/maximize-tensorflow-performance-on-cpu-considerations-and-recommendations-for-inference. 

As your issue is about ResNet 50 under windows,  if possible, could you please post your question to the Intel tensorflow github   https://github.com/Intel-tensorflow/tensorflow. so we can investigate there. 

Here we may focus on Bert issue, which was a known issue. and we fixed it in TF 
 master code with new PR, like https://github.com/tensorflow/tensorflow/pull/31623.  So for fix the Bert performance issue,  it required  rebuild TF with source code  instead of install from exist version. 
  
For rebuild TF  with source code, please follow up the article: 
https://software.intel.com/en-us/articles/intel-optimization-for-tensorflow-installation-guide
git clone https://github.com/tensorflow/tensorflow
git checkout r1.14.0
…. 
#e.g
bazel build --config=mkl --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 --copt=-march=sandybridge --copt=-mtune=ivybridge --copt=-O3 //tensorflow/tools/pip_package:build_pip_package

**Comment by vpirogov** [comment]
Closing due to lack of activity. Feel free to reopen with new data.

**Comment by volker42maru** [comment]
Hi @yinghu5 ,
you mentioned that performance on BERT is better with tf1.14 compiled with a newer version of MKL-DNN. Could you tell me, how I specify the MKL-DNN version when compiling TF from source?

**Comment by yinghu5** [comment]
@volker42maru  The latest TF source code will include download new version of MKL-DNN for example 0.20.3. so you can git clone it and build from source directly. 

build guide: https://software.intel.com/en-us/articles/intel-optimization-for-tensorflow-installation-guide

**Comment by volker42maru** [comment]
Thank you for the explanation @yinghu5.

I built tf1.15 with mkl-dnn 0.20.3 from source. However, inference is still around 2.5x slower compared to vanilla tf. I am running BERT on an 'older' CPU (Intel Xeon E5-2637 v4), so idk if that might be an issue.

I compiled tf for the AVX2 instruction set, because this CPU does not support AVX512.
